File: src/main/java/jb/files/FileReaderNWaySameValidation.java
package jb.files;

import de.bwaldvogel.liblinear.InvalidInputDataException;
import de.bwaldvogel.liblinear.Problem;
import jb.config.Opts;
import jb.data.Dataset;
import jb.data.clfobj.ClfObject;
import jb.data.clfobj.ClfObjectDoubleSorted;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class FileReaderNWaySameValidation implements FileHelper {

    @Override
    public Dataset readFile(Opts opts) throws IOException, InvalidInputDataException {

        Problem problem = Problem.readFromFile(new File(opts.getFilePath()), opts.getBias());
        ClfObject[] clfObjectDoubleSorteds = new ClfObjectDoubleSorted[problem.l];
        for (int i = 0; i < problem.l; i++) {
            double[] x = new double[problem.n];
            for (int j = 0; j < problem.n; j++) {
                x[j] = problem.x[i][j].getValue();
            }
            clfObjectDoubleSorteds[i] = new ClfObjectDoubleSorted(x, (int) problem.y[i]);
        }
        Arrays.sort(clfObjectDoubleSorteds);

        int numberOfSubsets = opts.getNumberOfBaseClassifiers() + 2;
        int countOfSubset = problem.l / (numberOfSubsets);
        List<jb.data.Problem> problems = new ArrayList<>();
        for (int i = 0; i < numberOfSubsets; i++) {
            double[][] x = new double[countOfSubset][problem.n];
            int[] y = new int[countOfSubset];
            for (int j = 0; j < countOfSubset; j++) {
                x[j] = clfObjectDoubleSorteds[j * numberOfSubsets + i].getX();
                y[j] = clfObjectDoubleSorteds[j * numberOfSubsets + i].getY();
            }
            problems.add(new jb.data.Problem(x, y));
        }
        return new Dataset(problems, clfObjectDoubleSorteds[0].getX()[0], clfObjectDoubleSorteds[clfObjectDoubleSorteds.length - 1].getX()[0]);
    }

}

File: src/main/java/jb/files/FileReaderNWaySameValidation.kt
package jb.files

import de.bwaldvogel.liblinear.InvalidInputDataException
import de.bwaldvogel.liblinear.Problem
import jb.config.Opts
import jb.data.Dataset
import jb.data.clfobj.ClfObject
import jb.data.clfobj.ClfObjectDoubleSorted

import java.io.File
import java.io.IOException
import java.util.ArrayList
import java.util.Arrays

class FileReaderNWaySameValidation : FileHelper {

    @Throws(IOException::class, InvalidInputDataException::class)
    override fun readFile(opts: Opts): Dataset {

        val problem = Problem.readFromFile(File(opts.filePath!!), opts.bias)
        val clfObjectDoubleSorteds = arrayOfNulls<ClfObjectDoubleSorted>(problem.l)
        for (i in 0 until problem.l) {
            val x = DoubleArray(problem.n)
            for (j in 0 until problem.n) {
                x[j] = problem.x[i][j].value
            }
            clfObjectDoubleSorteds[i] = ClfObjectDoubleSorted(x, problem.y[i].toInt())
        }
        Arrays.sort(clfObjectDoubleSorteds)

        val numberOfSubsets = opts.numberOfBaseClassifiers + 2
        val countOfSubset = problem.l / numberOfSubsets
        val problems = ArrayList<jb.data.Problem>()
        for (i in 0 until numberOfSubsets) {
            val x = Array(countOfSubset) { DoubleArray(problem.n) }
            val y = IntArray(countOfSubset)
            for (j in 0 until countOfSubset) {
                x[j] = clfObjectDoubleSorteds[j * numberOfSubsets + i]!!.x
                y[j] = clfObjectDoubleSorteds[j * numberOfSubsets + i]!!.y
            }
            problems.add(jb.data.Problem(x, y))
        }
        return Dataset(problems, clfObjectDoubleSorteds[0]!!.x[0], clfObjectDoubleSorteds[clfObjectDoubleSorteds.size - 1]!!.x[0])
    }

}
